packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts

```diff
@@ -0,0 +1,231 @@
+}
+
+export interface IJobQueue {
+	get length(): number
+	clear(): void
+	endsWith(jobId: string): boolean
+	getLastJobIdByImportance(importance: JobImportance): string | undefined
+	enqueue<T, V>(
+		job: Job<T, void, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T>
+	removeJob(jobId: string): void
+}
+
+export class JobQueue extends EventEmitter implements IJobQueue {
+	private queue: JobDescription<any>[] = []
+	private currentJob: JobDescription<any> | null = null
+	private currentJobChain: JobDescription<any>[] = []
+	private currentJobTimeout: NodeJS.Timeout | null = null
+
+	constructor(private name: string, private logger: Logger) {
+		super()
+	}
+
+	get length(): number {
+		return this.queue.length
+	}
+
+	clear(): void {
+		if (this.currentJob) {
+			this.abortJob(this.currentJob)
+		} else {
+			if (this.currentJobChain.length) {
+				this.currentJob = this.currentJobChain[this.currentJobChain.length - 1]
+
+				Promise.allSettled(this.currentJobChain.map(async (job) => job.job.cleanup()))
+					.then(this.handleCleanupResults)
+					.finally(this.advanceJob)
+				this.removeChainedJobsFromTheFront()
+			}
+		}
+	}
+
+	enqueue<T, U>(
+		job: Job<T, void, U>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T> {
+		this.queue.push({ id, job, chained: false, timeout, importance, abortController: new AbortController() })
+		this.run()
+		return this as IChainableJobQueue<T>
+	}
+
+	chain<T, U, V>(
+		job: Job<T, U, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T> {
+		this.queue.push({ id, job, chained: true, timeout, importance, abortController: new AbortController() })
+		return this as IChainableJobQueue<T>
+	}
+
+	end<X>(onFulfilled?: FulfilledCallback<X>, onRejected?: RejectedCallback, onTimeout?: TimeoutCallback): void {
+		const lastJob = this.queue[this.queue.length - 1]
+		if (!lastJob) return
+		lastJob.onFulfilled = onFulfilled
+		lastJob.onRejected = onRejected
+		lastJob.onTimeout = onTimeout
+	}
+
+	getLastJobIdByImportance(importance: JobImportance): string | undefined {
+		for (let i = this.queue.length - 1; i >= 0; --i) {
+			if (this.queue[i].importance === importance) {
+				return this.queue[i].id
+			}
+		}
+		if (this.currentJob?.importance === importance) {
+			return this.currentJob.id
+		}
+		return undefined
+	}
+
+	removeJob(jobId: string): void {
+		this.queue = this.queue.filter((job) => job.id !== jobId && !job.chained)
+	}
+
+	endsWith(jobId: string): boolean {
+		const lastJob = this.queue[this.queue.length - 1] ?? this.currentJob
+		return lastJob?.id === jobId
+	}
+
+	private run(): void {
+		setImmediate(() => {
+			if (this.currentJob) {
+				return
+			}
+			if (this.queue.length === 0 && this.currentJobChain.length === 0) {
+				this.emit('done')
+			}
+			this.currentJob = this.queue.shift() ?? null
```

Comment:
I am pretty sure the Nullish coalescing operator operator is redundant.

If the result of `this.queue.shift()` is null, then currently null is assinged to `this.currentJob` with or without the Nullish coalescing operator.

If a non-null value is to be assigned in the case of a null result from `this.queue.shift()`, then it is needed, but as is, should not be the case.

I am unsure if the Nullish coalescing operator has any effect incase `this.queue.shift()` returns `undefined`, if possibly.

Reply:
The result of Array.shif() is never null. It returns the removed element or `undefined`

Reply:
Ahh i see, Then the Nullish coalescing turns the undefined into a null.
I just checked how `.shift()` works in a sandbox, don't know why i didn't before writing it.

That said i think i found a bug in 'codesandbox.io'. It seems to have trouble with understanding a Nullish coalescing operator, seeing/expecting it as two ternary operators.